utils/android/adb/commands.py

```diff
@@ -0,0 +1,151 @@
+
+    1. 'adb shell' can only take input up to a certain length, so it fails for
+       long executable names or when a large amount of arguments are passed to
+       the executable. This function attempts to limit the size of any string
+       passed to 'adb shell'.
+    2. 'adb shell' ignores the exit code of any command it runs. This function
+       therefore uses its own mechanisms to determine whether the executable
+       had a successful exit code when run on device.
+
+    If 'rerun_failures' is specified, this function re-runs failing tests,
+    printing their output to stdout.
+    """
+    # We'll be running the executable in a temporary directory in
+    # /data/local/tmp. `adb shell` has trouble with commands that
+    # exceed a certain length, so to err on the safe side we only
+    # use the first 10 characters of the UUID.
+    uuid_dir = '{}/{}'.format(DEVICE_TEMP_DIR, str(uuid.uuid4())[:10])
+    shell(['mkdir', '-p', uuid_dir])
+
+    # `adb` can only handle commands under a certain length. No matter what the
+    # original executable's name, on device we call it `__executable`.
+    executable = '{}/__executable'.format(uuid_dir)
+    push(executable_path, executable)
+
+    # When running the executable on the device, we need to pass it the same
+    # arguments, as well as specify the correct LD_LIBRARY_PATH. Save these
+    # to a file we can easily call multiple times.
+    executable_with_args = '{}/__executable_with_args'.format(uuid_dir)
+    _create_executable_on_device(
+        executable_with_args,
+        'LD_LIBRARY_PATH={uuid_dir}:{tmp_dir} '
+        '{executable} {executable_arguments}'.format(
+            uuid_dir=uuid_dir,
+            tmp_dir=DEVICE_TEMP_DIR,
+            executable=executable,
+            executable_arguments=' '.join(executable_arguments)))
+
+    # Write the output from the test executable to a file named '__stdout', and
+    # if the test executable succeeds, write 'SUCCEEDED' to a file
+    # named '__succeeded'. We do this because `adb shell` does not report
+    # the exit code of the command it executes on the device, so instead we
+    # check the '__succeeded' file for our string.
+    executable_stdout = '{}/__stdout'.format(uuid_dir)
+    succeeded_token = 'SUCCEEDED'
+    executable_succeeded = '{}/__succeeded'.format(uuid_dir)
+    executable_piped = '{}/__executable_piped'.format(uuid_dir)
+    _create_executable_on_device(
+        executable_piped,
+        '{executable_with_args} > {executable_stdout} && '
+        'echo "{succeeded_token}" > {executable_succeeded}'.format(
+            executable_with_args=executable_with_args,
+            executable_stdout=executable_stdout,
+            succeeded_token=succeeded_token,
+            executable_succeeded=executable_succeeded))
+
+    # We've pushed everything we need to the device.
+    # Now execute the wrapper script.
+    shell([executable_piped])
```

Comment:
> Should I add some logic to `command_upload_stdlib` to clean the device each time?

This would make sense.

> Should I modify this to always remove temporary directories?

I think so.  In fact, you can remove all of them in `command_upload_stdlib` to prepare a device for a new test run.  I don't think you need to remove files during a single testsuite run, the total size of all executables is not that large (less than 300 Mb IIRC).

But there also needs to be logic that will clean up leftover processes, which might be stuck consuming 100% CPU.  One simple way to do that would be to reboot the device from `command_upload_stdlib`, possibly checking some conditions first.